File: tests/fixtures/usage-graph-cpp/consumer.cpp
#include "service.h"

namespace example {

class Consumer {
public:
    void viaInstance() {
        Service s;
        s.run();
    }

    void viaPointer(Service* p) {
        p->run();
    }

    void viaStatic() {
        Service::helper();
    }

    void callsLocal() {
        local();
    }

    void recurse() {
        recurse();
    }

    void wrongReceiver(Other* o) {
        o->run();
    }

    Service* makeService() {
        return new Service();
    }

    void viaFree() {
        freeHelper();
    }

    void local() {}
};

} // namespace example

File: tests/fixtures/usage-graph-scala/Consumer.scala
package example

class Consumer {
  def viaInstance(): Int = {
    val s = new Service()
    s.run()
  }

  def viaParam(svc: Service): Int = {
    svc.run()
  }

  def viaObject(): Int = {
    Helpers.help()
  }

  def callsLocal(): Int = {
    local()
  }

  def local(): Int = 7

  def makeService(): Service = {
    new Service()
  }

  def wrongReceiver(other: Consumer): Int = {
    other.run()
  }

  def recurse(): Int = {
    recurse()
  }
}
